native/meeseeks_html5ever_nif/src/lib.rs

```diff
@@ -126,20 +125,15 @@ enum ParserType {
     XmlDocument,
 }
 
-fn parse<'a>(parser_type: ParserType, env: Env<'a>, args: &[Term<'a>]) -> NifResult<Term<'a>> {
+fn parse<'a>(parser_type: ParserType, env: Env<'a>, document: Binary) -> NifResult<Term<'a>> {
```

Comment:
This change seems to work fine, tests pass, but it drops the previous panic, for what I guess will be a rustler specific error?

Reply:
Can you explain to me the change in signature?

Is that a change in Rustler v0.22? Can you point me to the appropriate documentation there?

Reply:
It was changed, but to be honest, the documentation is either hidden or not there, as the only way I figured out about this is by looking at others who had already updated.

The error without it is the one I mentioned before:

```
error[E0277]: the trait bound `&[rustler::Term<'a>]: Decoder<'_>` is not satisfied
   --> src/lib.rs:177:1
    |
177 | #[rustler::nif(schedule = "DirtyCpu")]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Decoder<'_>` is not implemented for `&[rustler::Term<'a>]`
    |
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `&[rustler::Term<'a>]: Decoder<'_>` is not satisfied
   --> src/lib.rs:182:1
    |
182 | #[rustler::nif(schedule = "DirtyCpu")]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Decoder<'_>` is not implemented for `&[rustler::Term<'a>]`
    |
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```

I'm *very* new to Rust, and don't really have a clue as to why this change is needed. I'm not sure if there's a way to keep `Term` and get it to work with `Decoder`, but if there is then maybe we can do this instead. No preference on my end here, just trying to get it to compile (which it does 😄)

Reply:
That something called "args", the first element of which was the document that needed parsing, is getting replaced by the document that is being parsed makes me think there is going to be a mismatch there.

I will need time to review Rustler's changes and figure out what's going on and how best to address that issue.

Reply:
[This](https://github.com/rusterlium/rustler/pull/264/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) (and some other files in that PR) are the only references that change this as well, but I couldn't really find it in the changelog, but I may have missed it.

Reply:
> I will need time to review Rustler's changes and figure out what's going on and how best to address that issue.

No worries, let me know if I can do anything. I'll be testing this in our app today, and maybe in production after that. If all goes well, I'll mark this as ready for review, so you can review it whenever you're ready.

Thanks :)

Reply:
> I'll be testing this in our app today,

That went a bit faster than I expected. All our tests still pass, and I manually checked all of our uses of Meeseeks and they are working, too. I'll report back when it's been running in production.

Reply:
Ok, I confirmed that your approach (replacing the slice of terms that was `args` with the actual arg, in this case `document`) is the correct one.

That was the approach suggested [here](https://github.com/rusterlium/rustler/issues/310#issuecomment-624343982) by @evnu.

Reply:
> I'm very new to Rust, and don't really have a clue as to why this change is needed.

The rustler changelog actually mentions this, but it is possibly too opaque.

> NIFs can be derived from regular functions, if the arguments implement Decoder and the return type implements Encoder:

Before the upcoming `rustler`, `args` was a slice of `Term` which needed to be decoded. Each entry in `args` corresponds with one argument passed into the function. Now with rustler v0.22, the arguments are spelled out explicitly, just like in a regular function.

Do you have a suggestion to make this more obvious for the users of rustler? Where would you expect that information?

Reply:
@evnu maybe some examples of the before and after could help. I figured this out by comparing libraries that had already upgraded from those that had not. Maybe `UPGRADING.md` file?

Maybe the changelog entry was fine, but like I said, I'm not familiar with rustler or rust, so someone else may have read `Encoder/Decoder` and knew immediately that those sliced params (?) would be replaced with it.

Reply:
@evnu I saw the example in the changelog under the "NIFs can be derived from regular functions..." note but I wasn't sure if it related to this because the signatures were so different: 

```rust
add<'a>(env: Env<'a>, args: &[Term<'a>]) -> Result<Term<'a>, Error>
```
vs

```rust
fn add(a: i64, b: i64) -> i64
```

An example that just showed args being replaced by each individual arg would have been clearer.

Reply:
> Maybe UPGRADING.md file?

That's a good idea, thanks!